pyembed/src/importer.rs

```diff
@@ -900,7 +904,219 @@ impl OxidizedFinder {
+            };
+            py.eval(&code, None, None).map_or_else(
+                |err_creating_exc| err_creating_exc,
+                |exc| PyErr::from_instance(py, exc),
+            )
+        })
+    }
+
+    fn path_hook_impl(&self, py: Python, path: PyObject) -> PyResult<_PathEntryFinder> {
+        // Compute _PathEntryFinder::package
+        let pkg = {
+            let current_exe = self.exe_realpath(py)?;
+            let not_exe_err = || {
+                let msg = format!(
+                    "path {} does not begin in \'{}\' (if you're sure it does, check Python's file-system encoding)", &path, current_exe.display());
+                let mut err = PyErr::new::<ImportError, _>(py, msg);
+                let exc = err.instance(py);
+                if let Err(setattr_err) = exc.setattr(py, "path", path.clone_ref(py)) {
+                    setattr_err
+                } else {
+                    err
+                }
+            };
+            let pbuf = pyobject_to_pathbuf(py, path.clone_ref(py))?;
+            let mut p: &std::path::Path = pbuf.as_ref();
+            // path could point "inside" current_exe. If we can't canonicalize,
+            // strip the last component and check again. (zipimporter does
+            // somethings similar)
+            let mut abs_p = p.canonicalize();
+            // Count the number of components at the end of `path` will be in pkg
+            let mut parts: usize = 0;
+            while abs_p.is_err() {
+                p = match p.parent() {
+                    Some(parent) => {
+                        parts += 1;
+                        parent
+                    }
+                    None => {
+                        return Err(not_exe_err());
+                    }
+                };
+                abs_p = p.canonicalize();
+            }
+            let abs_p = abs_p.unwrap();
+            if abs_p != current_exe {
+                return Err(not_exe_err());
+            }
+            // Extract the part of `path` after current_exe
+            // FIXME: This takes two allocations, but it shouldn't even use one.
+            let tail = pbuf
+                .iter()
+                .rev()
+                .take(parts)
+                .collect::<std::path::PathBuf>()
+                .iter()
+                .rev()
+                .collect::<std::path::PathBuf>();
+            _PathEntryFinder::parse_path_to_pkg(py, &tail)?
+        };
+        let finder = OxidizedFinder::create_instance(py, Arc::clone(self.state(py)))?;
+        // sys.path_hooks is only read after importlib._bootstrap_internal has
+        // been setup, either in that module or after initialization by
+        // PyImport_GetImporter checking for the existence of config->run_filename
+        // has been setup, so it's safe to import os now.
+        let paths = PyList::new(py, &[py.import("os")?.call(py, "fspath", (path,), None)?]);
```

Comment:
Since this is normalizing an argument value and it should be safe to call at the top of the function, could you please move this there?

While you are here, possibly add a comment saying why this is necessary (I know the reason, but it wasn't obvious when I first saw this code).

Reply:
3ae1efb2677b3e01ef3f2fbe6784e2059c017680